onnxruntime/core/providers/nnapi/nnapi_execution_provider.cc

```diff
@@ -0,0 +1,353 @@
+#include "core/framework/compute_capability.h"
+#include "core/session/onnxruntime_cxx_api.h"
+#include "core/session/inference_session.h"
+#include "core/graph/model.h"
+#include "dnnlibrary/ModelBuilder.h"
+#include "dnnlibrary/OnnxReader.h"
+#include "tools/onnx2daq/OnnxConverter.h"
+
+namespace onnxruntime {
+
+constexpr const char* NNAPI = "Nnapi";
+
+NnapiExecutionProvider::NnapiExecutionProvider()
+    : IExecutionProvider{onnxruntime::kNnapiExecutionProvider} {
+  DeviceAllocatorRegistrationInfo device_info{OrtMemTypeDefault,
+                                              [](int) { return std::make_unique<CPUAllocator>(
+                                                            std::make_unique<OrtAllocatorInfo>(NNAPI,
+                                                                                               OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemTypeDefault)); },
+                                              std::numeric_limits<size_t>::max()};
+  InsertAllocator(CreateAllocator(device_info));
+
+  DeviceAllocatorRegistrationInfo cpu_allocator_info({OrtMemTypeCPUOutput,
+                                                      [](int) { return std::make_unique<CPUAllocator>(std::make_unique<OrtAllocatorInfo>(NNAPI, OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemTypeCPUOutput)); },
+                                                      std::numeric_limits<size_t>::max()});
+
+  InsertAllocator(CreateAllocator(cpu_allocator_info));
+}
+
+NnapiExecutionProvider::~NnapiExecutionProvider() {}
+
+Status NnapiExecutionProvider::CopyTensor(const Tensor& src, Tensor& dst) const {
+  if (!(strcmp(src.Location().name, NNAPI) == 0 && strcmp(dst.Location().name, CPU) == 0) &&
+      !(strcmp(src.Location().name, CPU) == 0 && strcmp(dst.Location().name, NNAPI) == 0) &&
+      !(strcmp(src.Location().name, NNAPI) == 0 && strcmp(dst.Location().name, NNAPI) == 0)) {
+    ORT_NOT_IMPLEMENTED(src.Location().name, " copy to ", dst.Location().name, " is not implemented");
+  }
+
+  // Todo: Copy for now. May optimize later to avoid copy.
+  size_t bytes = src.DataType()->Size() * src.Shape().Size();
+  const void* src_data = src.DataRaw();
+  void* dst_data = dst.MutableDataRaw();
+  memcpy(dst_data, src_data, bytes);
+
+  return Status::OK();
+}
+
+std::vector<std::vector<int>> NnapiExecutionProvider::GetSupportedNodes(const ONNX_NAMESPACE::ModelProto& model_proto) const {
+  dnn::OnnxConverter converter;
+  return converter.GetSupportedNodes(model_proto);
+}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+NnapiExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
+                                      const std::vector<const KernelRegistry*>& /*kernel_registries*/) const {
+  // This method is based on that of TRT EP
+  // Construct modelproto from graph
+  onnxruntime::Model model(graph.Name(), true, ModelMetaData(), IOnnxRuntimeOpSchemaRegistryList(), graph.DomainToVersionMap());
```

Comment:
it's necessary to add a "ToProto()" interface in GraphViewer to avoid duplicating this kind of logic in every provider. Let me know if you want to do the refactoring or not, if no, I'll do it later (may after your check-in).

Reply:
I'd like to submit a sperate PR to refactor it :)